charms/istio-pilot/src/charm.py

```diff
@@ -95,17 +124,36 @@ def send_info(self, event):
             )
 
     def handle_ingress(self, event):
-        ingress = self.interfaces['ingress']
+        gateway_address = self._get_gateway_address()
+        if not gateway_address:
+            self.unit.status = WaitingStatus("Waiting for gateway address")
+            event.defer()
+            return
+        else:
+            self.unit.status = ActiveStatus()
 
+        ingress = self.interfaces['ingress']
         if ingress:
-            routes = ingress.get_data().items()
+            # Filter out data we sent back.
+            routes = {
+                (rel, app): route
+                for (rel, app), route in sorted(
+                    ingress.get_data().items(), key=lambda tup: tup[0][0].id
+                )
+                if app != self.app
+            }
         else:
-            routes = []
+            routes = {}
+
+        if isinstance(event, RelationBrokenEvent):
+            # The app-level data is still visible on a broken relation, but we
+            # shouldn't be keeping the VirtualService for that related app.
+            del routes[(event.relation, event.app)]
```

Comment:
Has this been tested?  I cherry picked this in #49 but when testing it encountered this error when I broke a relation:
```
Traceback (most recent call last):
  File "./src/charm.py", line 266, in <module>
    main(Operator)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/main.py", line 406, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/main.py", line 140, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/framework.py", line 278, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/framework.py", line 722, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-istio-pilot-0/charm/venv/ops/framework.py", line 767, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 143, in handle_ingress
    del routes[(event.relation, event.app)]
TypeError: 'dict_items' object does not support item deletion
unit-istio-pilot-0: 11:10:39 ERROR juju.worker.uniter.operation hook "ingress-relation-broken" (via hook dispatching script: dispatch) failed: exit status 1
unit-istio-pilot-0: 11:10:39 INFO juju.worker.uniter awaiting error resolution for "relation-broken" hook
```

I may have broken something myself in the process though.  I'm going to add a unit test and debug this in #49

Reply:
nevermind, I did not also cherry pick the updated way routes was constructed (right above this), so I had a dict_items object instead of a dict. I think this was my own mistake and yours works (although a relation-broken test would still help)

Reply:
Yeah, that looks like a cherry-pick issue.  The relation remove is covered by [the cleanup portion](https://github.com/canonical/istio-operators/pull/43/files#diff-f602add18718f057def3bdfbfb4f0f923ae6cfe123a82b2ff8f639de063fe288R166-R167), and that error implies that you must be missing [this bit of the change](https://github.com/canonical/istio-operators/pull/43/files#diff-d3a4e684bd1a3a1822df5ab7605cfce00c97cd7b52f44ed770526c4f4f80eacdR136-R146) which makes `routes` an actual `dict` rather than the `dict_items` list-like.